### PR TITLE
Telegram support

### PR DESCRIPTION
**Pull request by 0xCA**
# Telegram support
image
> Telegram userid settings and config send button
> Userid setting is moved to a separate collapsible section

## Reason
- In short, convenience and time saving
- Mainly intended for quick effortless distribution, especially for large servers and mass config updates

## Details
- You can assign one user id or group id to each client config.
- Each telegram userid can be assigned to multiple configs, so one user can have multiple configs.
- You can search by userid on the clients page.
- There is a configurable _global_ flood wait. It applies to both manual config push and a user request. 1h default.
- I used dependency injection to avoid circular dependencies with util.
- There is an in-memory cache of `tg userid -> client ids` correspondence to avoid heavy db usage.
- Pre-generated QR codes are kinda low res for telegram. So I generate 512x512 images each time they are needed. 

## Config request
image
> Send any message to the bot to get **all** the configs associated with your userid
- If config requests are disabled, bot module will not listen to updates. You can have a separate server handling the incoming messages without conflicts.
- If you have multiple configs assigned to you, they are sent with a 2s delay between them, to avoid hitting bot API limits.
- If there are no configs assigned, flood wait is still applied to avoid bot overload.

## Configuration
| Variable                      | Description                                                                                                                                                  | Default                            |
| ----------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------------------------- |
| `TELEGRAM_TOKEN`              | Telegram bot token for distributing configs to clients                                                                                                       | N/A                                |
| `TELEGRAM_ALLOW_CONF_REQUEST` | Allow users to get configs from the bot by sending a message                                                                                                 | `false`                            |
| `TELEGRAM_FLOOD_WAIT`         | Time in minutes before the next conf request is processed                                                                                                    | `60`                               |

None of this configuration is saved to the database. You can change it by restarting the wgui.

# Conflicts
I tried to minimize conflicts for this branch, but they are inevitable,
#### The advised merge order:
1. #470
2. #481
3. #483
4. this

If you intend to merge all of the above, you can simplify this by merging https://github.com/0xCA/wireguard-ui/tree/umaster, where all the conflicts are already resolved.

## Future
As this repo is currently inactive, it's inconvenient to continue development from the stale master. Further development (if any) is going to be in my repo. Feel free to check and contribute.

### CODE REVIEW

**Comment by nebulosa2007** [comment]
Also thought about adding Telegram ID in clients settings :) For now I use Email field for filling Telegram Id and just parse json files with my own tgbot. Nothing special - it parses wgui configs and send .conf files with generated QR-code to user according his TgID.

This setup is more convenient than yours, because conf files send to user when he wants, but not when you press the button in wgui admin interface (besides user must add tgbot to contacts before, that you could send him an anything).

**Comment by X-UIP** [comment]
How can update Wireguard-UI to get this ability?

**Comment by X-UIP** [comment]
> @X-UIP There is no public image built from my branch. You should build it yourself. Hint: you will need to update Go version in Dockerfile. Use `golang:1.20-alpine3.17` as build and `alpine:3.17` as runtime.

excuse me! if you can create a learning! because I don't know exactly what you say! I can create and run, Wireguard UI with a docker file and special image, but your work i don't know yours!

**Comment by ngoduykhanh** [comment]
Hey @0xCA, thank you for another contribution here. I guess I should touch this PR after solving #481. I've just left a comment there.
Although I am a Telegram user, I don't really see how this feature would be helpful in my daily use case. However, I am happy to merge if somebody else needs it. Let me know if you still want to continue with your contribution here or if you want to stop and keep it in your folks.

**Comment by ngoduykhanh** [comment]
@0xCA fair enough. I merged the other PR. could you please resolve this PR conflict as well?

File: telegram/bot.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+type TgBotInitDependencies struct {
+	DB                             store.IStore
+	SendRequestedConfigsToTelegram SendRequestedConfigsToTelegram
+}
+
+var (
+	TelegramToken            string
+	TelegramAllowConfRequest bool
+	TelegramFloodWait        int
```
Comment (ngoduykhanh) [inline telegram/bot.go:23]:
```suggestion
	Token            string
	AllowConfRequest bool
	FloodWait        int
```

The variable name shouldn't start with package name.

File: telegram/bot.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	LogLevel                 log.Lvl
+
+	TgBot      *echotron.API
+	TgBotMutex sync.RWMutex
```
Comment (ngoduykhanh) [inline telegram/bot.go:27]:
```suggestion
	Bot      *echotron.API
	BotMutex sync.RWMutex
```

I think it is a better naming

**Review by ngoduykhanh — CHANGES_REQUESTED** [review]
Found an issue. When the `TELEGRAM_ALLOW_CONF_REQUEST` is not set and it has `false` value by default, I cannot send the Telegram user the config. 

telegram-req-error

When `TELEGRAM_ALLOW_CONF_REQUEST=true`, everything is fine.

**Comment by ngoduykhanh** [comment]
Perfect! I am merging the PR. Thank you for addressing all the issues.

**Comment by PulseDiver** [comment]
Where should I add these settings - TELEGRAM_TOKEN, TELEGRAM_ALLOW_CONF_REQUEST and TELEGRAM_FLOOD_WAIT?

**Pull request merged**